ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_15/EventQueryServiceIT.scala

```diff
@@ -0,0 +1,269 @@
+    }
+  })
+
+  test(
+    "TXEventsByContractIdNotVisible",
+    "No events are returned for a non-visible contract id",
+    allocate(TwoParties),
+  )(implicit ec => { case Participants(Participant(ledger, party, notTheSubmittingParty)) =>
+    for {
+      tx <- ledger.submitAndWaitForTransaction(
+        ledger.submitAndWaitRequest(party, Dummy(party).create.command)
+      )
+      expected = assertDefined(
+        tx.transaction.flatMap(_.events.flatMap(_.event.created).headOption),
+        "Expected a created event",
+      )
+      events <- ledger.getEventsByContractId(
+        GetEventsByContractIdRequest(
+          expected.contractId,
+          immutable.Seq(notTheSubmittingParty.unwrap),
+        )
+      )
+    } yield {
+      assertIsEmpty(Seq(events.createEvent, events.archiveEvent).flatten[GeneratedMessage])
+    }
+  })
+
+  test(
+    "TXEventsByContractKeyBasic",
+    "Expose a visible create event by contract key",
+    allocate(SingleParty),
+  )(implicit ec => { case Participants(Participant(ledger, party)) =>
+    val someKey = "some key"
+    val key = makeTextKeyKey(party, someKey)
+
+    for {
+      tx <- ledger.submitAndWaitForTransaction(
+        ledger.submitAndWaitRequest(party, TextKey(party, someKey, Nil).create.command)
+      )
+      expected = assertDefined(
+        tx.transaction.flatMap(_.events.flatMap(_.event.created).headOption),
+        "Expected a created event",
+      )
+
+      events <- ledger.getEventsByContractKey(
+        GetEventsByContractKeyRequest(
+          contractKey = Some(key),
+          templateId = Some(TextKey.id.unwrap),
+          requestingParties = Tag.unsubst(immutable.Seq(party)),
+        )
+      )
+    } yield {
+      val actual = assertDefined(events.createEvent, "Expected a created event")
+      assertEquals("Looked up event should match the transaction event", actual, expected)
+    }
+  })
+
+  test(
+    "TXArchiveEventByContractKeyBasic",
+    "Expose a visible archive event by contract key",
+    allocate(SingleParty),
+  )(implicit ec => { case Participants(Participant(ledger, party)) =>
+    val someKey = "some archive key"
+    val key = makeTextKeyKey(party, someKey)
+
+    for {
+      cId <- ledger.create(party, TextKey(party, someKey, Nil))
+      tx <- ledger.submitAndWaitForTransaction(
+        ledger.submitAndWaitRequest(party, cId.exerciseTextKeyChoice().command)
+      )
+      expected = assertDefined(
+        tx.transaction.flatMap(_.events.flatMap(_.event.archived).headOption),
+        "Expected an archived event",
+      )
+      events <- ledger.getEventsByContractKey(
+        GetEventsByContractKeyRequest(
+          contractKey = Some(key),
+          templateId = Some(TextKey.id.unwrap),
+          requestingParties = Tag.unsubst(immutable.Seq(party)),
+        )
+      )
+    } yield {
+      val actual = assertDefined(events.archiveEvent, "Expected a archived event")
```

Comment:
Same as above. I can imagine a future bug where populating archived obliterates the created